### DISCUSSION TITLE
MQTT Authentication?

### DISCUSSION THREAD

**Discussion by konistehrad** [chrivers/bifrost#5] [Q&A] [answered]
Excited to start taking a look at this package, thanks for whipping it up! I made some cursory patches to diyHue to support MQTT device selection, but this looks significantly more robust. 

My MQTT server, however, is currently password protected and doesn't allow anonymous access. I see that the `Z2mServer` struct only has one element: `url`. Should the password and username be included in standard URL format there, like `ws://user:pass@localhost:8080`?

**Comment by chrivers** [comment, marked as answer]
Hello @konistehrad - and welcome to bifrost :)

Thanks you for saying bifrost looks more robust. That was a big focus during initial development!

In my own case, I'm securing the network z2m is on, instead of z2m itself, so I haven't encountered exactly this problem before.

I didn't implement the low-level websocket logic from scratch - I use a crate called `tungstenite`, which handles that for me. It's that same crate that parses the url, so... uhm... maybe that would work? 😅

Can you try it, and see if it does?

Also notice that bifrost is actually connecting to z2m not over mqtt (ironically), but rather over the (somewhat undocumented) websocket z2m has, which the z2m ui uses.

I'd *imagine* that authentication on z2m also includes this websocket, but I'm not actually sure?

Can you test it with the `ws://user:pass@localhost:8080` scheme, and tell me if it works? Then we'll figure out what we can do, if it doesn't

**Reply by konistehrad** [reply]
Ah, I see so I'm connecting to z2m directly, not to MQTT, got it. In that case I just opened a PR to allow use of the `wss` scheme in the case that Z2M is behind TLS. Extending that, I use the `auth_token` to secure the API endpoint; I'll be evaluating adding support for that in the coming days. Thanks again!

**Reply by chrivers** [reply]
That sounds good - let me know if I can help!

I think `wss://` should be fairly easy to add support for. I guess the `auth_token` is sent as a header, or some kind of websocket equivalent?

If you can find some kind of documentation (or API example) of how to use the `auth_token`, I'm sure we can find a way to enable it :+1:

**Reply by konistehrad** [reply]
`auth_token` is apparently sent as part of the query string to the `/api` endpoint: `/api?token=XXX` like that. It doesn't appear to be documented.

**Reply by chrivers** [reply]
If you have discord, feel free to stop by so we can have a chat about it :)

https://discord.gg/YvBKjHBJpA

**Reply by Flavien** [reply]
What was the solution to connect to a Z2M install that has a password?

**Reply by Flavien** [reply]
Ok found it: `ws://ip:port/api?token=yourpassword`

**Reply by chrivers** [reply]
Oh, very nice! I was looking at it for a little bit, and I didn't find the correct incantation.

I'll try include this in the README in the next iteration :+1:

**Comment by chrivers** [comment]
Update: This has been included in the README for a while now, and Bifrost even detects if it's missing, and gives a helpful warning if you don't specify a token :+1:
